onboarding/mlsys/2-autodiff.md

```diff
@@ -0,0 +1,399 @@
+
+## TensorFlow
+
+While we do have more PyTorch users than TensorFlow/Keras users, but we develop
+features and field support questions from both frameworks.  We also maintain a
+Keras-based feature in the upstream Horovod project.  So it is important to be
+familiar with both.
+
+### Q7
+
+Train a model using momentum again, this time in TensorFlow.  As before, your
+solution should use low-level Tensor manipulations rather than high-level APIs.
+You might start reading about [tf.Variable](
+https://www.tensorflow.org/guide/variable) and [tf.GradientTape](
+https://www.tensorflow.org/guide/autodiff).
+
+Note that tf.Variable and tf.GradientTape will come up in OSS support, whenever
+users subclass `keras.models.Model` or `keras.metrics.Metric`.
+
+Here's a template to get you started:
+
+```python
+import tensorflow as tf
+from tensorflow.keras import layers, losses, models, optimizers
+
+LR = 0.01
+MOMENTUM = 0.9
+
+# same bogus model as with the pytorch example
+model = models.Sequential()
+model.add(layers.Dense(5, activation=None, use_bias=False, input_shape=(5,)))
+model.add(layers.Dense(5, activation=None, use_bias=False))
+opt = optimizers.SGD(learning_rate=LR, momentum=MOMENTUM)
+model.compile(
+    optimizers.SGD(learning_rate=LR, momentum=MOMENTUM),
+    losses.mean_squared_error,
+)
+
+# Your code here:
+# - copy initial weights
+# - define my_train_batch()
+
+# test your momentum implementation against tensorflow's
+for i in range(100):
+    data = tf.random.uniform((1, 5))
+    tfloss = model.train_on_batch(data, data)
+    myloss = my_train_iter(data)
+    assert abs(tfloss - myloss) < 0.000001, (
+        f"i = {i}, tfloss = {tfloss}, myloss = {myloss}"
+    )
+```
+
+## Optimizers and Large Models
+
+As you have seen, optimizer state is often at least as large as model state,
+even for fairly optimizers strategies like momentum.  This effect is compounded
+with other advanced features like mixed-precision training, where mixed
+precision, where even though model parameters require less memory each, the
+per-parameter optimizer state memory requirement actually grows.
+
+As a result, training large models often means dealing with very large
+optimizer states.  A popular strategy is the [ZeRO Memory Optimization](
+https://arxiv.org/abs/1910.02054).
+
+### Q8
+
+Go read the [ZeRO Memory Optimization](https://arxiv.org/abs/1910.02054) paper.
+Read at least the intro and section 5.1 about Optimizer State Partitioning
+($P_{os}$).  Then take the momentum training loop you wrote  and write a training
+loop that simulates momentum-based training with Optimizer State Partitioning
+across two workers.  Feel free to do it in PyTorch or TensorFlow.
+
+No need for actual distributed mechanics; just do it inside a single process.
+But do make sure your code demonstrates how the optimizer state can be sharded
+so each worker only needs to store half of the momentum state.
+
+## Summary
+
+Upon completion of this exercise, you have learned:
```

Comment:
try either one of these 

Upon completion of this exercise, you will have learned:

or

Congratulations for completing this exercise! In this exercise, you learned: